pkg/agent/openflow/pipeline.go

```diff
@@ -2704,3 +3033,297 @@ func (sl conjunctiveActionsInOrder) Less(i, j int) bool {
+	}
+	for _, ipProtocol := range f.ipProtocols {
+		flows = append(flows,
+			// This generates the default flow to match the first packet of Service connection.
+			PreRoutingClassifierTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchProtocol(ipProtocol).
+				Action().ResubmitToTables(targetTables...).
+				Done(),
+		)
+	}
+
+	return flows
+}
+
+// Feature: featureService
+// Stage: stageRouting
+// Tables: L3ForwardingTable
+// l3FwdFlowsToExternal generates the flows to forward the packets of Service connection to external network.
+func (f *featureService) l3FwdFlowsToExternal() []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	if f.connectUplinkToBridge {
+		flows = append(flows,
+			// Regardless of whether the traffic mode is encap or noEncap, the corresponding traffic models are:
+			//   - Local Pod --> External / Local Node (via OVS gateway port)
+			//      * -rpl+trk, FromLocalRegMark, FromLocalCTMark, RewriteMACRegMark, ServiceCTMark (If Egress is disabled)
+			//   - External / Local Gateway (via OVS gateway port) --> External / Local Node (via OVS gateway port)
+			//      * -rpl+trk, FromGatewayRegMark, FromGatewayCTMark, RewriteMACRegMark, ServiceCTMark, HairpinCTMark
+			// When the traffic mode is noEncap, the unique corresponding traffic models are:
+			//   - Local Pod --> Remote Antrea Pod (via OVS gateway port)
+			//      * -rpl+trk, FromLocalRegMark, FromLocalCTMark, RewriteMACRegMark, ServiceCTMark
+			//   - Local Gateway --> Remote Antrea Pod (via OVS gateway port)
+			//      * -rpl+trk, FromGatewayRegMark, FromGatewayCTMark, RewriteMACRegMark, ServiceCTMark, HairpinCTMark
+			L3ForwardingTable.ofTable.BuildFlow(priorityLow).
+				Cookie(cookieID).
+				MatchRegMark(RewriteMACRegMark).
+				MatchRegMark(NotAntreaFlexibleIPAMRegMark).
+				MatchCTMark(ServiceCTMark).
+				Action().SetDstMAC(f.gatewayMAC).
+				Action().LoadRegMark(ToGatewayRegMark).
+				Action().NextTable().
+				Done(),
+			// This generates the flow to match packets sourced from Antrea IPAM Pod and destined for external network.
+			// The traffic mode is only noEncap, and corresponding traffic models are:
+			//   - Antrea Pod --> Remote Antrea Pod / External (via OVS uplink port)
+			//     * -rpl+trk, FromLocalRegMark, FromLocalCTMark, RewriteMACRegMark, ServiceCTMark, AntreaFlexibleIPAMRegMark
+			L3ForwardingTable.ofTable.BuildFlow(priorityLow).
+				Cookie(cookieID).
+				MatchRegMark(RewriteMACRegMark).
+				MatchRegMark(AntreaFlexibleIPAMRegMark).
+				MatchCTMark(ServiceCTMark).
+				Action().ResubmitToTables(L3DecTTLTable.GetID()).
+				Done(),
+		)
+	} else {
+		// This generates the flow to forward the packets sourced from local Pods and destined for external network.
+		flows = append(flows,
+			L3ForwardingTable.ofTable.BuildFlow(priorityLow).
+				Cookie(cookieID).
+				MatchRegMark(RewriteMACRegMark).
```

Comment:
Do you think adding a match with from local reg is more accurate?

Reply:
Not just packets from local Pods. Packets sourced from gateway and destined for external network are also matched by the flow.